apps/namadillo/src/hooks/useIbcTransaction.tsx

```diff
@@ -0,0 +1,142 @@
+import { useMemo } from "react";
+import {
+  Address,
+  AddressWithAssetAndAmount,
+  ChainRegistryEntry,
+  GasConfig,
+  TransferTransactionData,
+} from "types";
+
+type useIbcTransactionProps = {
+  sourceAddress?: string;
+  registry?: ChainRegistryEntry;
+  sourceChannel?: string;
+  shielded?: boolean;
+  destinationChannel?: Address;
+  selectedAsset?: AddressWithAssetAndAmount;
+};
+
+import { Keplr, Window as KeplrWindow } from "@keplr-wallet/types";
+import { QueryStatus } from "@tanstack/query-core";
+import { ibcTransferAtom } from "atoms/integrations";
+import BigNumber from "bignumber.js";
+import { getIbcGasConfig } from "integrations/utils";
+import { useAtomValue } from "jotai";
+
+type useIbcTransactionOutput = {
+  transferToNamada: (
+    destinationAddress: string,
+    displayAmount: BigNumber
+  ) => Promise<TransferTransactionData>;
+  transferStatus: "idle" | QueryStatus;
+  gasConfig: GasConfig | undefined;
+  assertValid: () => void;
+};
+
+export const useIbcTransaction = ({
+  registry,
+  sourceAddress,
+  selectedAsset,
+  sourceChannel,
+  shielded,
+  destinationChannel,
+}: useIbcTransactionProps): useIbcTransactionOutput => {
+  const performIbcTransfer = useAtomValue(ibcTransferAtom);
+
+  const gasConfig = useMemo(() => {
+    if (typeof registry !== "undefined") {
+      return getIbcGasConfig(registry);
+    }
+    return undefined;
+  }, [registry]);
+
+  const getWallet = (): Keplr => {
+    const wallet = (window as KeplrWindow).keplr;
+    if (typeof wallet === "undefined") {
+      throw new Error("No Keplr instance");
+    }
+    return wallet;
+  };
+
+  const assertValid = (): void => {
+    if (typeof sourceAddress === "undefined") {
+      throw new Error("Source address is not defined");
+    }
+
+    if (!selectedAsset) {
+      throw new Error("No asset is selected");
+    }
+
+    if (!registry) {
+      throw new Error("Invalid chain");
+    }
+
+    if (!sourceChannel) {
+      throw new Error("Invalid IBC source channel");
+    }
+
+    if (shielded && !destinationChannel) {
+      throw new Error("Invalid IBC destination channel");
+    }
+
+    if (typeof gasConfig === "undefined") {
+      throw new Error("No transaction fee is set");
+    }
+  };
+
+  const transferToNamada = async (
+    destinationAddress: Address,
+    displayAmount: BigNumber
+  ): Promise<TransferTransactionData> => {
+    assertValid();
```

Comment:
This implementation is breaking the type check and requiring you to add the `!` on the variables below.

To avoid this issue, we could update it to be like:
```tsx
invariant(sourceAddress, "Source address is not defined")
invariant(selectedAsset, "No asset is selected")
invariant(registry, "Invalid chain")
invariant(sourceChannel, "Invalid IBC source channel")
invariant(shielded && !destinationChannel, "Invalid IBC destination channel")
invariant(gasConfig, "No transaction fee is set")
```

I understand it could be too verbose, but it's quite often that we catch an border case error because of these `!` 

Another alternative is creating a type that holds all params as required and build this object with some type check on another function, like `buildIbcTxParams: (sourceAddress?: string .... ) => IbcTxParams`